File: SundeeFundee/Domain/Calculations/WeightCalculations.swift
import Foundation

// MARK: - Weight Calculations

enum SessionResult { case first, success, failure }

enum WeightCalculations {

    static func roundToNearestFive(_ value: Double) -> Double {
        (value / 5.0).rounded() * 5.0
    }

    // MARK: - Equipment-specific weight snapping

    /// Available dumbbell weights in pounds (the actual inventory).
    static let availableDumbbellWeightsLb: [Double] = [10, 15, 20, 25, 35, 40, 50, 70]

    /// Available kettlebell weights in pounds.
    static let availableKettlebellWeightsLb: [Double] = [15, 25, 32, 53, 70]

    /// Snaps a barbell total to the nearest weight that can actually be loaded
    /// using standard plates (45, 25, 15, 10, 5, 2.5 lb), accounting for bar weight.
    /// Plates go on in 2.5 lb increments per side, so valid totals are bar + multiples of 5 lb.
    /// - Parameters:
    ///   - lb: The target weight in pounds.
    ///   - barLb: Bar weight in pounds (45 for men's, 35 for women's).
    /// - Returns: The nearest loadable total in pounds.
    static func snapBarbellWeightLb(_ lb: Double, barLb: Double = 45.0) -> Double {
        // Each side gets plates in 2.5 lb increments, so total plate weight moves in 5 lb steps.
        // Snap to nearest 5 lb total above the bar weight.
        let plateLoadLb = max(0, lb - barLb)
        let snappedPlateLoad = (plateLoadLb / 5.0).rounded() * 5.0
        let snappedTotalLb = barLb + snappedPlateLoad
        return snappedTotalLb
    }

    /// Snaps a weight in lb to the nearest available dumbbell weight.
    static func snapDumbbellWeightLb(_ lb: Double) -> Double {
        availableDumbbellWeightsLb.min(by: { abs($0 - lb) < abs($1 - lb) }) ?? lb
    }

    /// Snaps a weight in lb to the nearest available kettlebell weight.
    static func snapKettlebellWeightLb(_ lb: Double) -> Double {
        availableKettlebellWeightsLb.min(by: { abs($0 - lb) < abs($1 - lb) }) ?? lb
    }

    static func calculateTargetWeight(oneRepMax: Double, percentage: Double) -> Double {
        roundToNearestFive(oneRepMax * percentage)
    }

    static func getNextRecommendedWeight(
        currentWeight: Double,
        result: SessionResult,
        oneRepMax: Double
    ) -> Double {
        switch result {
        case .first:
            return roundToNearestFive(oneRepMax * 0.7)
        case .success:
            return roundToNearestFive(currentWeight + 5)
        case .failure:
            let floor = roundToNearestFive(oneRepMax * 0.5)
            return max(roundToNearestFive(currentWeight - 5), floor)
        }
    }

    static func wasSetSuccessful(
        actualReps: Int,
        prescribedReps: Int,
        actualWeight: Double,
        prescribedWeight: Double?
    ) -> Bool {
        let repsOk = actualReps >= prescribedReps
        let weightOk: Bool
        if let prescribedWeight {
            weightOk = actualWeight >= prescribedWeight
        } else {
            weightOk = true
        }
        return repsOk && weightOk
    }

    static func isPersonalRecord(weight: Double, previousMax: Double) -> Bool {
        weight > previousMax
    }

    static func calculateVolumeLoad(weight: Double, reps: Int, sets: Int) -> Double {
        weight * Double(reps) * Double(sets)
    }

    static func detectPlateau(weights: [Double]) -> Bool {
        guard weights.count >= 3 else { return false }
        let last3 = Array(weights.suffix(3))
        return (last3.max()! - last3.min()!) < 5
    }
}

// MARK: - Epley 1RM Formula

enum EpleyFormula {
    /// Estimates 1RM from a submaximal lift.
    /// Formula: weight × (1 + reps / 30)
    /// Only valid for reps > 1 (for reps == 1, the lift itself is the 1RM).
    static func estimated1RM(weight: Double, reps: Int) -> Double {
        guard reps > 1 else { return weight }
        return weight * (1.0 + Double(reps) / 30.0)
    }

    /// Returns true if the new estimated 1RM is a PR vs the stored max.
    static func isPR(newEstimate: Double, currentMax: Double?) -> Bool {
        guard let current = currentMax else { return true }
        return newEstimate > current
    }
}

File: src/domain/calculations/weight-calculations.ts
/**
 * Weight Calculations
 * Ported from SundeeFundee/Domain/Calculations/WeightCalculations.swift
 *
 * All functions are pure (no mutation of inputs).
 * Uses undefined (not null) for absent values.
 */

import type { WeightUnit } from '../types';

// ---------------------------------------------------------------------------
// Equipment inventories
// ---------------------------------------------------------------------------

/** Available dumbbell weights in pounds (standard gym inventory) */
export const AVAILABLE_DUMBBELL_WEIGHTS_LB: readonly number[] = [10, 15, 20, 25, 35, 40, 50, 70];

/** Available kettlebell weights in pounds */
export const AVAILABLE_KETTLEBELL_WEIGHTS_LB: readonly number[] = [15, 25, 32, 53, 70];

// ---------------------------------------------------------------------------
// Rounding helpers
// ---------------------------------------------------------------------------

/**
 * Round a value to the nearest multiple of 5.
 * Uses Swift-parity rounding (round half away from zero).
 */
export function roundToNearestFive(value: number): number {
  return Math.round(value / 5.0) * 5.0;
}

// ---------------------------------------------------------------------------
// Barbell weight snapping
// ---------------------------------------------------------------------------

/**
 * Snap a barbell total to the nearest loadable weight in pounds.
 * Plate load moves in 2.5 lb increments per side (5 lb total steps).
 *
 * @param lb    - Target weight in pounds
 * @param barLb - Bar weight in pounds (default: 45 for men's bar)
 */
export function snapBarbellWeightLb(lb: number, barLb = 45.0): number {
  const plateLoadLb = Math.max(0, lb - barLb);
  const snappedPlateLoad = Math.round(plateLoadLb / 5.0) * 5.0;
  return barLb + snappedPlateLoad;
}

/**
 * Snap a barbell total to the nearest loadable weight in kilograms.
 * Plate load moves in 2.5 kg increments per side (5 kg total steps).
 *
 * @param kg    - Target weight in kilograms
 * @param barKg - Bar weight in kilograms (default: ~20.41 kg, 45 lb standard bar)
 */
export function snapBarbellWeightKg(kg: number, barKg = 45.0 / 2.2046226218): number {
  const plateLoadKg = Math.max(0, kg - barKg);
  const snappedPlateLoad = Math.round(plateLoadKg / 5.0) * 5.0;
  return barKg + snappedPlateLoad;
}

// ---------------------------------------------------------------------------
// Dumbbell / Kettlebell snapping
// ---------------------------------------------------------------------------

/**
 * Snap a weight in lb to the nearest available dumbbell weight.
 */
export function snapDumbbellWeightLb(lb: number): number {
  return (
    AVAILABLE_DUMBBELL_WEIGHTS_LB.reduce((best, w) =>
      Math.abs(w - lb) < Math.abs(best - lb) ? w : best
    ) ?? lb
  );
}

/**
 * Snap a weight in lb to the nearest available kettlebell weight.
 */
export function snapKettlebellWeightLb(lb: number): number {
  return (
    AVAILABLE_KETTLEBELL_WEIGHTS_LB.reduce((best, w) =>
      Math.abs(w - lb) < Math.abs(best - lb) ? w : best
    ) ?? lb
  );
}

// ---------------------------------------------------------------------------
// Target weight calculation
// ---------------------------------------------------------------------------

/**
 * Calculate target weight as a percentage of 1RM, rounded to nearest 5.
 */
export function calculateTargetWeight(oneRepMax: number, percentage: number): number {
  return roundToNearestFive(oneRepMax * percentage);
}

// ---------------------------------------------------------------------------
// Progress utilities
// ---------------------------------------------------------------------------

/**
 * Returns true if the given weight exceeds the previous max.
 * Note: named isPersonalRecord (not isPR) to avoid conflict with epley-formula export.
 */
export function isPersonalRecord(weight: number, previousMax: number): boolean {
  return weight > previousMax;
}

/**
 * Calculate total volume load for a set/rep/weight combination.
 */
export function calculateVolumeLoad(weight: number, reps: number, sets: number): number {
  return weight * reps * sets;
}

/**
 * Returns true if the last 3 weights in the array differ by less than 5 (plateau signal).
 */
export function detectPlateau(weights: number[]): boolean {
  if (weights.length < 3) return false;
  const last3 = weights.slice(-3);
  const maxVal = Math.max(...last3);
  const minVal = Math.min(...last3);
  return maxVal - minVal < 5;
}

// Re-export WeightUnit for convenience
export type { WeightUnit };
